File: 2025-1/sesion-de-clase-09/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/inf30/softprog/rrhh/daoimpl/AreaDAOImpl.java
package pe.edu.pucp.inf30.softprog.rrhh.daoimpl;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import pe.edu.pucp.inf30.softprog.rrhh.dao.IAreaDAO;
import pe.edu.pucp.inf30.softprog.rrhh.model.Area;

public class AreaDAOImpl extends BaseDAOImpl<Area> implements IAreaDAO {
    @Override
    protected CallableStatement comandoInsertar(Connection conn, Area area) throws SQLException {
        String sql = "{CALL insertarArea(?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_nombre", area.getNombre());
        cmd.setBoolean("p_activo", area.isActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        return cmd;
    }

    @Override
    protected CallableStatement comandoModificar(Connection conn, Area area) throws SQLException {
        String sql = "{CALL modificarArea(?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_nombre", area.getNombre());
        cmd.setBoolean("p_activo", area.isActivo());
        cmd.setInt("p_id", area.getId());
        return cmd;
    }

    @Override
    protected CallableStatement comandoEliminar(Connection conn, int id) throws SQLException {
        String sql = "{CALL eliminarArea(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        return cmd;
    }

    @Override
    protected CallableStatement comandoBuscar(Connection conn, int id) throws SQLException {
        String sql = "{CALL buscarAreaPorId(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        return cmd;
    }

    @Override
    protected CallableStatement comandoListar(Connection conn) throws SQLException {
        String sql = "{CALL listarAreas()}";
        CallableStatement cmd = conn.prepareCall(sql);
        return cmd;
    }
    
    @Override
    protected Area mapearModelo(ResultSet rs) throws SQLException {
        Area area = new Area();
        area.setId(rs.getInt("id"));
        area.setNombre(rs.getString("nombre"));
        area.setActivo(rs.getBoolean("activo"));
        return area;
    }
}

File: 2025-1/sesion-de-clase-13 - Solucion/SoftProgPersistencia/DAOImpl/RRHH/AreaDAOImpl.cs
using System.Data;
using System.Data.Common;
using PUCP.Edu.Pe.SoftProg.Modelo.RRHH;
using PUCP.Edu.Pe.SoftProg.Persistencia.DAO.RRHH;

namespace PUCP.Edu.Pe.SoftProg.Persistencia.DAOImpl.RRHH {
    public class AreaDAOImpl : BaseDAOImpl<Area>, IAreaDAO {
        protected override DbCommand CommandoInsertar(DbConnection conn, Area area) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarArea";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, 
                "@p_nombre", DbType.String, area.Nombre);
            this.AgregarParametroEntrada(cmd, 
                "@p_activo", DbType.Boolean, area.IsActive);
            this.AgregarParametroSalida(cmd, 
                "@p_id", DbType.Int32);

            return cmd;
        }

        protected override DbCommand CommandoModificar(DbConnection conn, 
            Area area) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarArea";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_nombre", 
                DbType.String, area.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_activo", 
                DbType.Boolean, area.IsActive);
            this.AgregarParametroEntrada(cmd, "@p_id", 
                DbType.Int32, area.Id);

            return cmd;
        }

        protected override DbCommand CommandoEliminar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarArea";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand CommandoBuscar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarAreaPorId";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;
            
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand CommandoListar(DbConnection conn) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarAreas";
            cmd.CommandType = System.Data.CommandType.StoredProcedure;
            return cmd;
        }

        protected override Area mapearModelo(DbDataReader reader) {
            return new Area {
                Id = (int)reader["id"], 
                Nombre = reader["nombre"].ToString(),
                IsActive = (bool)reader["activo"],
            };
        }
    }
}
